sqlmesh/core/snapshot/evaluator.py

```diff
@@ -1661,6 +1663,93 @@ def get_custom_materialization_type(name: str) -> t.Type[CustomMaterialization]:
     return strategy_type
 
 
+class EngineManagedStrategy(PromotableStrategy):
```

Comment:
Right, so, I think I have figured out what to do here.

On `create()`, because the DeployabilityIndex always returns False to allow creating self-referential models, we just create the dev table and carry on. The call to create the prod table is a no-op because we dont yet know if it will be deployable.

On `insert()`, the DeployabilityIndex works as expected. If the current snapshot is deployable, we call `create_managed_table`. Otherwise, we replace the dev table with the preview data. We only do this for the "first insert" because managed models are not incremental.

This approach means that promotion works as expected with no need for a table creation side-effect.

Reply:
This sounds really good 👍

Reply:
Actually, this won't work if the "MANAGED" model snapshot IS deployable and has downstream dependencies that are part of the same plan. The downstream models will attempt to create their own tables referencing the deployable version of the "MANAGED" model and it will be missing at that point.

Reply:
What you probably need to do is to check whether the snapshot is deployable **before** we modify the deployability index to make the current snapshot non-deployable. Eg. add something like this:
```
is_snapshot_deployable = deployability_index.is_deployable(snapshot)
```

Reply:
Of course :disappointed: , I was so focused on upstream models and if they were forward-only or not that I completely forgot about downstream models.

I've implemented your suggestion to pass the snapshot deployability to `create()` since we cant rely on the DeployabilityIndex there.

This changes the logic to:

- on `create()`, if both the table and snapshot are deployable, we create a managed table. Otherwise, we fallback to the superclass logic which just creates a normal table
- on `insert()`, we populate the table only if the snapshot is NOT deployable (which would have resulted in a normal table being created in `create()`). If the snapshot IS deployable, then it's a no-op because the data would have been populated in `create()`.

Reply:
There's now an extra case added to `insert()`. If a snapshot that was not deployable at creation (and thus would have been created as a normal table) suddenly becomes deployable - then we "upgrade" it to a managed table by dropping it and recreating it as a managed table